sdk/python/kfp/dsl/_component_bridge.py

```diff
@@ -18,31 +18,86 @@
-        if isinstance(argument_value, dsl.PipelineParam):
+        if isinstance(argument_value, dsl.ContainerOp):
+            raise TypeError('ContainerOp object was passed to component as an input argument. Pass a single output instead.')
+        elif isinstance(argument_value, dsl.PipelineParam):
             input_type = component_spec._inputs_dict[input_name].type
             reference_type = argument_value.param_type
             dsl.types.verify_type_compatibility(reference_type, input_type, 'Incompatible argument passed to the input "{}" of component "{}": '.format(input_name, component_spec.name))
 
             arguments[input_name] = str(argument_value)
-        if isinstance(argument_value, dsl.ContainerOp):
-            raise TypeError('ContainerOp object was passed to component as an input argument. Pass a single output instead.')
 
+            # TODO: support more artifact types
+            if input_type.lower() == 'gcspath':
+                pipeline_task_spec.inputs.artifacts[input_name].producer_task = argument_value.op_name or 'need_an_importer_node'
+                pipeline_task_spec.inputs.artifacts[input_name].output_artifact_key = argument_value.name
+            elif input_type.lower() in ['string', 'str', 'integer', 'int', 'double', 'float']:
+                pipeline_task_spec.inputs.parameters[input_name].runtime_value.runtime_parameter = argument_value.name
+            else:
+                raise NotImplementedError('Unsupported input type: "{}"'.format(input_type))
+        elif isinstance(argument_value, str):
+            pipeline_task_spec.inputs.parameters[input_name].runtime_value.constant_value.string_value = argument_value
+        elif isinstance(argument_value, int):
+            pipeline_task_spec.inputs.parameters[input_name].runtime_value.constant_value.int_value = argument_value
+        elif isinstance(argument_value, float):
+            pipeline_task_spec.inputs.parameters[input_name].runtime_value.constant_value.double_value = argument_value
+
+    if component_spec.outputs:
+        for output in component_spec.outputs:
+            if output.type.lower() == 'gcspath':
+                pipeline_task_spec.outputs.artifacts[output.name].artifact_type.schema_title = "mlpipeline.GenericArtifact"
+            elif output.type.lower() in ['string', 'str', 'integer', 'int', 'double', 'float']:
+                # TODO: support output parameter
+                pass
+
+    def _sanitize_file_name(name):
+        import re
+        return re.sub('[^-_.0-9a-zA-Z]+', '_', name)
+
+    def _ir_input_path_generator(port_name: str) -> str:
+        return "{{$.inputs.artifacts['" + _sanitize_file_name(port_name) + "'].uri}}"
+
+
+    def _ir_output_path_generator(port_name: str) -> str:
+        return "{{$.outputs.artifacts['" + _sanitize_file_name(port_name) + "'].uri}}"
+
+
+    def  _ir_input_value_generator(value, type_name: str, input_name: str) -> str:
```

Comment:
I do not think this is needed. Just pass the placeholder strings as arguments:

```
argument_placeholders = {input_spec.name, "{{$.inputs.parameters['" + input_spec.name + "']}}" for input_spec in component_spec.inputs or []}
```

Reply:
This likely doesn't work with the current component I/O design where we have the separation of `{{$.inputs.parameters[...]}}` and `{{$.inputs.artifacts[...]}}`.

With this `_ir_input_value_generator`, I want it to only apply on argument of `InputValuePlaceholder` (replacing `argument_serializer`), while argument of `InputPathPlaceholder` is handled differently.

https://github.com/kubeflow/pipelines/blob/06bf42998ecacecac2fe50e11c43c8a64922289b/sdk/python/kfp/components/_components.py#L400-L406

Reply:
You're right. Now I see how this trick works: if the value is already a string than `serialize_value` would simply return it as-is. Will revert the unnecessary changes.